lib/puppet/type/file/ensure.rb

```diff
@@ -175,6 +183,19 @@ def retrieve
     end
 
     def sync
+      if self.should == :absent && !@resource.bucket.nil?
+        # Make a dummy content property so that Puppet
+        # will generate a content object, which we can
+        # use to the checksum of the current file.
+        @resource[:content] = :absent
+        # Make sure we're using the same hash as the
+        # bucket, so that our log is actually useful.
+        # This may need to change when PUP-4329 is fixed.
+        @resource[:checksum] = Puppet[:digest_algorithm].to_sym
```

Comment:
This means the `removed file` message won't match up with the `defined content` message if `Puppet[:digest_algorithm]` (global config) differs from `@resource[:checksum]` (resource property), but this was my suggested approach to fulfill the desired behavior of mapping to the Filebucket contents until PUP-4329 is fixed. The checksums will still be correct and self-consistent, as they include the checksum type in the message.